### PR TITLE
[Safety Rules] Avoid exporting the consensus private key

### PR DESCRIPTION
**Pull request by JoshLind**
## Motivation

This PR updates Libra Safety Rules (LSR) to support two modes of operation: (i) exporting the consensus key from vault on each epoch change (which is the default today); and (ii) delegating the signing of data to vault directly, so that LSR doesn't need to export the key (the new mode of operation). The choice of which operational mode is defined by a new boolean in SafetyRulesConfig, "export_consensus_key", which this PR defaults to false.

There are several caveats to this change:
1. If vault is expected to do the signing, on each epoch change, to be sure vault actually holds the consensus private key for the current key version, we sign a piece of test data. If this fails, we immediately know that the key doesn't exist. This prevents us from having to wait until our first sign attempt to realize we don't have the key.
2. There will be a slight performance hit from delegating signatures to vault. We'll need to confirm this exact number in canary/cluster test (old results showed 5% - 15%, but we've since optimized the storage path, this may change).

This PR offers 3 commits to achieve delegated consensus key signatures:
1. First, we replace the ValidatorSigner in LSR with a ConfigurableValidatorSigner and add a new configuration option to SafetyRulesConfig.
2. Next, we update the existing unit tests to support the different operational modes.
3. Finally, we add two new smoke tests to ensure that both operational modes continue to work.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

All tests pass.

## Related PRs

This PR is based on an old proof of concept PR that we previously closed out (https://github.com/libra/libra/pull/6004)

### CODE REVIEW

**Comment by davidiw** [comment]
Is this ready for review? Not clear looking at the latest canary.

**Comment by JoshLind** [comment]
> Is this ready for review? Not clear looking at the latest canary.

Technically this is ready for review (so if you have early feedback, go for it). However, it's failing on CT because we've renamed the consensus key in vault (so I still need to look into either updating CT to avoid looking for the old key, or find a meaningful work around where we can support both keys).

File: consensus/safety-rules/src/safety_rules.rs [hunk 1]
```diff
@@ -249,26 +260,31 @@ impl SafetyRules {
...
-                        );
-
-                        self.validator_signer = None;
-                        error
-                    })?;
-
-                self.validator_signer = Some(ValidatorSigner::new(author, consensus_key));
+                let validator_handle = Some(ValidatorHandle::new(
+                    author,
+                    CONSENSUS_KEY.into(),
```
Comment (davidiw) [inline consensus/safety-rules/src/safety_rules.rs:267]: this is a const, why are we introducing it as a field?

File: config/global-constants/src/lib.rs [hunk 2]
```diff
@@ -9,7 +9,7 @@
 #![forbid(unsafe_code)]
 
 /// Definitions of global cryptographic keys (e.g., as held in secure storage)
-pub const CONSENSUS_KEY: &str = "consensus";
+pub const CONSENSUS_KEY: &str = "consensus_v2";
```
Comment (davidiw) [inline config/global-constants/src/lib.rs:12]: why are we introducing a new key?

**Review by davidiw — CHANGES_REQUESTED** [review]
At this point, I don't think we should be dictating behavioral changes in deployment. We can make recommendations and provide configuration to do so. We can then work with our partners to roll out improved security but everything should be incremental and reversible. TL;DR, keep names the same and rely on a config to dictate which behavior is desirable.

File: config/global-constants/src/lib.rs (hunk 2 above)
Comment (mgorven) [reply config/global-constants/src/lib.rs:12]: Let's handle this as a separate migration.

File: consensus/safety-rules/src/safety_rules.rs (hunk 1 above)
Comment (JoshLind) [reply consensus/safety-rules/src/safety_rules.rs:267]: Fair enough -- if we're always assuming a validator is identified by the CONSENSUS_KEY, I'll merge this into the ValidatorHandle 😄

File: config/global-constants/src/lib.rs (hunk 2 above)
Comment (JoshLind) [reply config/global-constants/src/lib.rs:12]: > why are we introducing a new key?

@davidiw, as discussed offline, this was because vault doesn't allow a key to be non-exportable after previously being exportable (see the PR description).

> Let's handle this as a separate migration.

Okay, thanks @mgorven, sounds good, I'll update the PR.

**Comment by JoshLind** [comment]
Thanks @davidiw!

> TL;DR, keep names the same and rely on a config to dictate which behavior is desirable.

As discussed offline, I'll update the PR to adopt a configurable approach. This not only makes sense for this controlled roll-outs and fail safes, but it makes sense for the TCB design going forward. I'll update the PR when it's ready for another look.

**Comment by JoshLind** [comment]
@davidiw, this PR is ready for another look, I've finally updated it to be configurable (although, we now configure this to be the default). Hopefully it's not too bad -- although the "configuration" aspect of it definitely introduced a lot more code 😕

File: consensus/safety-rules/src/configurable_validator_signer.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
```
Comment (davidiw) [inline consensus/safety-rules/src/configurable_validator_signer.rs:4]: this isn't required unless it isn't in main.rs and lib.rs files.

File: consensus/safety-rules/src/safety_rules.rs [hunk 4]
```diff
@@ -166,7 +179,7 @@ impl SafetyRules {
     }
 
     /// This verifies whether the author of one proposal is the validator signer
-    fn verify_author(&self, author: Option<Author>) -> Result<(), Error> {
+    fn verify_author(&mut self, author: Option<Author>) -> Result<(), Error> {
```
Comment (davidiw) [inline consensus/safety-rules/src/safety_rules.rs:182]: Why is this mutable? I think we should fish out whether or not the crypto APIs should be mutable for signing rather than polluting this. If, for whatever reason, they must be, we should hide that because in general signing shouldn't be mutable. Hiding via -- rwlock -- for example.

**Review by davidiw — COMMENTED** [review]
overall looks good, but before we pull in "tech debt" of mut, can we try to improve the hygiene of the code base and try to eliminate it?

File: consensus/safety-rules/src/configurable_validator_signer.rs (hunk 3 above)
Comment (JoshLind) [reply consensus/safety-rules/src/configurable_validator_signer.rs]: Ooh, thanks! Copy pasta 😄

File: consensus/safety-rules/src/safety_rules.rs (hunk 4 above)
Comment (JoshLind) [reply consensus/safety-rules/src/safety_rules.rs]: This is not strictly required (must have been legacy?), so it's been fixed here: https://github.com/libra/libra/pull/6762. I've rebased the PR and it looks better 😄

**Comment by JoshLind** [comment]
Thanks, @davidiw! 😄 

> overall looks good, but before we pull in "tech debt" of mut, can we try to improve the hygiene of the code base and try to eliminate it?

Done, hopefully it looks a little cleaner?

File: config/src/config/safety_rules_config.rs [hunk 5]
```diff
@@ -36,7 +37,8 @@ impl Default for SafetyRulesConfig {
             service: SafetyRulesService::Thread,
             test: None,
             verify_vote_proposal_signature: true,
-            // Default value of 30seconds for a timeout
+            export_consensus_key: false,
```
Comment (davidiw) [inline config/src/config/safety_rules_config.rs:40]: hmm....

@mgorven, is this good for you?

**Review by davidiw — APPROVED** [review]
wait for Michael's response and let's also ensure that perf isn't affected too much... if it is, you'll likely need to bother with that side of things :/

File: config/src/config/safety_rules_config.rs (hunk 5 above)
Comment (mgorven) [reply config/src/config/safety_rules_config.rs:40]: Defaulting to remote signing? Fine with me!

Comment (JoshLind) [reply config/src/config/safety_rules_config.rs:40]: Sounds good! I'll push it :smile
